File: infras/communication/Nm/Nm.c
/**
 * SSAS - Simple Smart Automotive Software
 * Copyright (C) 2024 Parai Wang <parai@foxmail.com>
 */
#ifdef USE_NM
/* ================================ [ INCLUDES  ] ============================================== */
#include "Nm.h"
#include "Nm_Priv.h"
#include "CanNm.h"
#include "OsekNm.h"
#include "UdpNm.h"
#include "ComM.h"
#include "Det.h"
/* ================================ [ MACROS    ] ============================================== */
#define NM_CONFIG (&Nm_Config)
/* ================================ [ TYPES     ] ============================================== */
/* ================================ [ DECLARES  ] ============================================== */
extern const Nm_ConfigType Nm_Config;
/* ================================ [ DATAS     ] ============================================== */
/* ================================ [ LOCALS    ] ============================================== */
/* ================================ [ FUNCTIONS ] ============================================== */
void Nm_Init(const Nm_ConfigType *ConfigPtr) {
}

Std_ReturnType Nm_PassiveStartUp(NetworkHandleType NetworkHandle) {
  Std_ReturnType ret = E_NOT_OK;
  const Nm_ChannelConfigType *config;

  DET_VALIDATE(NetworkHandle < NM_CONFIG->numOfChannels, 0x01, NM_E_INVALID_CHANNEL,
               return E_NOT_OK); /* @SWS_Nm_00488 */

  config = &NM_CONFIG->channelConfigs[NetworkHandle];
  switch (config->busType) {
#ifdef USE_CANNM
  case NM_BUSNM_CANNM:
    ret = CanNm_PassiveStartUp(config->handle);
    break;
#endif
#ifdef USE_OSEKNM
  case NM_BUSNM_OSEKNM:
    ret = OsekNm_NetworkRequest(config->handle);
    break;
#endif
  default:
    break;
  }

  return ret;
}

Std_ReturnType Nm_NetworkRequest(NetworkHandleType NetworkHandle) {
  Std_ReturnType ret = E_NOT_OK;
  const Nm_ChannelConfigType *config;

  DET_VALIDATE(NetworkHandle < NM_CONFIG->numOfChannels, 0x02, NM_E_INVALID_CHANNEL,
               return E_NOT_OK); /* @SWS_Nm_00489 */

  config = &NM_CONFIG->channelConfigs[NetworkHandle];
  switch (config->busType) {
#ifdef USE_CANNM
  case NM_BUSNM_CANNM:
    ret = CanNm_NetworkRequest(config->handle);
    break;
#endif
#ifdef USE_OSEKNM
  case NM_BUSNM_OSEKNM:
    ret = OsekNm_NetworkRequest(config->handle);
    break;
#endif
  default:
    break;
  }

  return ret;
}

Std_ReturnType Nm_NetworkRelease(NetworkHandleType NetworkHandle) {
  Std_ReturnType ret = E_NOT_OK;
  const Nm_ChannelConfigType *config;

  DET_VALIDATE(NetworkHandle < NM_CONFIG->numOfChannels, 0x03, NM_E_INVALID_CHANNEL,
               return E_NOT_OK); /* @SWS_Nm_00490 */

  config = &NM_CONFIG->channelConfigs[NetworkHandle];
  switch (config->busType) {
#ifdef USE_CANNM
  case NM_BUSNM_CANNM:
    ret = CanNm_NetworkRelease(config->handle);
    break;
#endif
#ifdef USE_OSEKNM
  case NM_BUSNM_OSEKNM:
    ret = OsekNm_NetworkRelease(config->handle);
    break;
#endif
  default:
    DET_VALIDATE(FALSE, 0x03, NM_E_PARAM_POINTER, ret = E_NOT_OK);
    break;
  }

  return ret;
}

Std_ReturnType Nm_GetState(NetworkHandleType nmNetworkHandle, Nm_StateType *nmStatePtr,
                           Nm_ModeType *nmModePtr) {
  Std_ReturnType ret = E_NOT_OK;
  const Nm_ChannelConfigType *config;

  DET_VALIDATE(nmNetworkHandle < NM_CONFIG->numOfChannels, 0x0E, NM_E_INVALID_CHANNEL,
               return E_NOT_OK); /* @SWS_Nm_00500 */

  config = &NM_CONFIG->channelConfigs[nmNetworkHandle];
  switch (config->busType) {
#ifdef USE_CANNM
  case NM_BUSNM_CANNM:
    ret = CanNm_GetState(config->handle, nmStatePtr, nmModePtr);
    break;
#endif
#ifdef USE_OSEKNM
  case NM_BUSNM_OSEKNM:
    ret = OsekNm_GetState(config->handle, nmModePtr);
    break;
#endif
  default:
    break;
  }

  return ret;
}

void Nm_MainFunction(void) {
}

void Nm_NetworkStartIndication(NetworkHandleType nmNetworkHandle) {
  const Nm_ChannelConfigType *config;
  (void)config;
  DET_VALIDATE(nmNetworkHandle < NM_CONFIG->numOfChannels, 0x11, NM_E_INVALID_CHANNEL, return);
  config = &NM_CONFIG->channelConfigs[nmNetworkHandle];
#ifdef USE_COMM /* @SWS_Nm_00155 */
  ComM_Nm_NetworkStartIndication(config->ComMHandle);
#endif
}

void Nm_NetworkMode(NetworkHandleType nmNetworkHandle) {
  const Nm_ChannelConfigType *config;
  (void)config;
  DET_VALIDATE(nmNetworkHandle < NM_CONFIG->numOfChannels, 0x12, NM_E_INVALID_CHANNEL, return);
  config = &NM_CONFIG->channelConfigs[nmNetworkHandle];
#ifdef USE_COMM /* @SWS_Nm_00158 */
  ComM_Nm_NetworkMode(config->ComMHandle);
#endif
}

void Nm_BusSleepMode(NetworkHandleType nmNetworkHandle) {
  const Nm_ChannelConfigType *config;
  (void)config;
  DET_VALIDATE(nmNetworkHandle < NM_CONFIG->numOfChannels, 0x14, NM_E_INVALID_CHANNEL, return);
  config = &NM_CONFIG->channelConfigs[nmNetworkHandle];
#ifdef USE_COMM /* @SWS_Nm_00163 */
  ComM_Nm_BusSleepMode(config->ComMHandle);
#endif
}

void Nm_PrepareBusSleepMode(NetworkHandleType nmNetworkHandle) {
  const Nm_ChannelConfigType *config;
  (void)config;
  DET_VALIDATE(nmNetworkHandle < NM_CONFIG->numOfChannels, 0x13, NM_E_INVALID_CHANNEL, return);
  config = &NM_CONFIG->channelConfigs[nmNetworkHandle];
#ifdef USE_COMM /* @SWS_Nm_00161 */
  ComM_Nm_PrepareBusSleepMode(config->ComMHandle);
#endif
}

void Nm_TxTimeoutException(NetworkHandleType nmNetworkHandle) {
}

void Nm_RepeatMessageIndication(NetworkHandleType nmNetworkHandle) {
}

void Nm_RemoteSleepIndication(NetworkHandleType nmNetworkHandle) {
}

void Nm_RemoteSleepCancellation(NetworkHandleType nmNetworkHandle) {
}

void Nm_CoordReadyToSleepIndication(NetworkHandleType nmChannelHandle) {
}

void Nm_CoordReadyToSleepCancellation(NetworkHandleType nmChannelHandle) {
}

void Nm_StateChangeNotification(NetworkHandleType nmNetworkHandle, Nm_StateType nmPreviousState,
                                Nm_StateType nmCurrentState) {
}

void Nm_CarWakeUpIndication(NetworkHandleType nmChannelHandle) {
}
#endif

File: tools/generator/Nm.py
# SSAS - Simple Smart Automotive Software
# Copyright (C) 2024 Parai Wang <parai@foxmail.com>

import pprint
import os
import json
from .helper import *

__all__ = ["Gen"]


def Gen_Nm(cfg, dir):
    H = open("%s/Nm_Cfg.h" % (dir), "w")
    GenHeader(H)
    H.write("#ifndef NM_CFG_H\n")
    H.write("#define NM_CFG_H\n")
    H.write("/* ================================ [ INCLUDES  ] ============================================== */\n")
    H.write("/* ================================ [ MACROS    ] ============================================== */\n")
    H.write("#ifndef NM_MAIN_FUNCTION_PERIOD\n")
    H.write("#define NM_MAIN_FUNCTION_PERIOD %s\n" % (cfg.get("MainFunctionPeriod", 10)))
    H.write("#endif\n")
    H.write("#define NM_CONVERT_MS_TO_MAIN_CYCLES(x) \\\n")
    H.write("  ((x + NM_MAIN_FUNCTION_PERIOD - 1) / NM_MAIN_FUNCTION_PERIOD)\n\n")
    H.write("/* ================================ [ TYPES     ] ============================================== */\n")
    H.write("/* ================================ [ DECLARES  ] ============================================== */\n")
    H.write("/* ================================ [ DATAS     ] ============================================== */\n")
    H.write("/* ================================ [ LOCALS    ] ============================================== */\n")
    H.write("/* ================================ [ FUNCTIONS ] ============================================== */\n")
    H.write("#endif /* NM_CFG_H */\n")
    H.close()

    C = open("%s/Nm_Cfg.c" % (dir), "w")
    GenHeader(C)
    C.write("/* ================================ [ INCLUDES  ] ============================================== */\n")
    C.write('#include "Nm.h"\n')
    C.write('#include "Nm_Cfg.h"\n')
    C.write('#include "Nm_Priv.h"\n')
    C.write("/* ================================ [ MACROS    ] ============================================== */\n")
    C.write("/* ================================ [ TYPES     ] ============================================== */\n")
    C.write("/* ================================ [ DECLARES  ] ============================================== */\n")
    C.write("/* ================================ [ DATAS     ] ============================================== */\n")
    C.write("static const Nm_ChannelConfigType Nm_ChannelConfigs[] = {\n")
    for idx, network in enumerate(cfg["networks"]):
        C.write("  {\n")
        C.write("    NM_BUSNM_%s, /* ImmediateNmCycleTime */\n" % (network.get("type", "CANNM")))
        C.write("    %s, /* handle */\n" % (idx))
        C.write("    %s, /* ComMHandle */\n" % (idx))
        C.write("  },\n")
    C.write("};\n\n")

    C.write("const Nm_ConfigType Nm_Config = {\n")
    C.write("  Nm_ChannelConfigs,\n")
    C.write("  ARRAY_SIZE(Nm_ChannelConfigs),\n")
    C.write("};\n")
    C.write("/* ================================ [ LOCALS    ] ============================================== */\n")
    C.write("/* ================================ [ FUNCTIONS ] ============================================== */\n")
    C.close()


def Gen(cfg):
    dir = os.path.join(os.path.dirname(cfg), "GEN")
    os.makedirs(dir, exist_ok=True)
    with open(cfg) as f:
        cfg = json.load(f)
    Gen_Nm(cfg, dir)
    return ["%s/Nm_Cfg.c" % (dir)]
